include/aws/iot/MqttRequestResponseClient.h

```diff
@@ -0,0 +1,378 @@
+                 */
+                SubscriptionEstablished = ARRSSET_SUBSCRIPTION_ESTABLISHED,
+
+                /**
+                 * The streaming operation has temporarily lost its subscription to its topic (filter)
+                 */
+                SubscriptionLost = ARRSSET_SUBSCRIPTION_LOST,
+
+                /**
+                 * The streaming operation has entered a terminal state where it has given up trying to subscribe
+                 * to its topic (filter).  This is always due to user error (bad topic filter or IoT Core permission
+                 * policy).
+                 */
+                SubscriptionHalted = ARRSSET_SUBSCRIPTION_HALTED,
+            };
+
+            /**
+             * An event that describes a change in subscription status for a streaming operation.
+             */
+            class AWS_CRT_CPP_API SubscriptionStatusEvent
+            {
+              public:
+                SubscriptionStatusEvent() : m_type(SubscriptionStatusEventType::SubscriptionEstablished), m_errorCode(0)
+                {
+                }
+                SubscriptionStatusEvent(const SubscriptionStatusEvent &rhs) = default;
+                SubscriptionStatusEvent(SubscriptionStatusEvent &&rhs) = default;
+                ~SubscriptionStatusEvent() = default;
+
+                SubscriptionStatusEvent &operator=(const SubscriptionStatusEvent &rhs) = default;
+                SubscriptionStatusEvent &operator=(SubscriptionStatusEvent &&rhs) = default;
+
+                SubscriptionStatusEvent &WithType(SubscriptionStatusEventType type)
+                {
+                    m_type = type;
+                    return *this;
+                }
+
+                SubscriptionStatusEvent &WithErrorCode(int errorCode)
+                {
+                    m_errorCode = errorCode;
+                    return *this;
+                }
+
+                SubscriptionStatusEventType GetType() const { return m_type; }
+                int GetErrorCode() const { return m_errorCode; }
+
+              private:
+                SubscriptionStatusEventType m_type;
+                int m_errorCode;
+            };
+
+            using SubscriptionStatusEventHandler = std::function<void(SubscriptionStatusEvent &&)>;
+
+            // @internal
+            class AWS_CRT_CPP_API IncomingPublishEvent
+            {
+              public:
+                IncomingPublishEvent() : m_payload() { AWS_ZERO_STRUCT(m_payload); }
+                IncomingPublishEvent(const IncomingPublishEvent &rhs) = default;
+                IncomingPublishEvent(IncomingPublishEvent &&rhs) = default;
+                ~IncomingPublishEvent() = default;
+
+                IncomingPublishEvent &operator=(const IncomingPublishEvent &rhs) = default;
+                IncomingPublishEvent &operator=(IncomingPublishEvent &&rhs) = default;
+
+                IncomingPublishEvent &WithPayload(Aws::Crt::ByteCursor payload)
```

Comment:
For mqtt5 client, we keep a copy of the buffer data. Not sure which way is a better approach though.

Reply:
Why make a copy unless it's necessary?